In [86]:
# Random Forest
import numpy as np
import pandas as pd
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error 
from sklearn.preprocessing import MinMaxScaler
from sklearn import svm
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [39]:
import sys
print(sys.version)

3.6.9 (default, Jul 17 2020, 12:50:27) 
[GCC 8.4.0]


In [48]:
pop = pd.read_csv('/content/drive/My Drive/project/seoul_contest/서울시 주민등록인구 (연령별동별) 통계.txt',header=None , sep='\t')

In [21]:
!wget -O mini.sh https://repo.anaconda.com/miniconda/Miniconda3-py38_4.8.2-Linux-x86_64.sh
!chmod +x mini.sh
!bash ./mini.sh -b -f -p /usr/local
!conda install -q -y --prefix /usr/local jupyter
!python -m ipykernel install --name "py38" --user

--2020-10-03 13:53:20--  https://repo.anaconda.com/miniconda/Miniconda3-py38_4.8.2-Linux-x86_64.sh
Resolving repo.anaconda.com (repo.anaconda.com)... 104.16.130.3, 104.16.131.3, 2606:4700::6810:8303, ...
Connecting to repo.anaconda.com (repo.anaconda.com)|104.16.130.3|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 89817099 (86M) [application/x-sh]
Saving to: ‘mini.sh’

mini.sh             100%[===================>]  85.66M   130MB/s    in 0.7s    

2020-10-03 13:53:21 (130 MB/s) - ‘mini.sh’ saved [89817099/89817099]

PREFIX=/usr/local
Unpacking payload ...
Solving environment: - \ | / - \ | / - \ | / - done

## Package Plan ##

  environment location: /usr/local

  added / updated specs:
    - _libgcc_mutex==0.1=main
    - asn1crypto==1.3.0=py38_0
    - ca-certificates==2020.1.1=0
    - certifi==2019.11.28=py38_0
    - cffi==1.14.0=py38h2e261b9_0
    - chardet==3.0.4=py38_1003
    - conda-package-handling==1.6.0=py38h7b6447c_0
    - 

In [22]:
!jupyter kernelspec list

Available kernels:
  py38       /root/.local/share/jupyter/kernels/py38
  ir         /usr/local/share/jupyter/kernels/ir
  python2    /usr/local/share/jupyter/kernels/python2
  python3    /usr/local/share/jupyter/kernels/python3
  swift      /usr/local/share/jupyter/kernels/swift


In [33]:
crime = pd.read_pickle('/content/drive/My Drive/project/seoul_contest/범죄율_주차구역_동별.pkl')
# age = pd.read_pickle('/content/drive/My Drive/project/seoul_contest/행정동별_5-14세_거주인구.pkl')
# welfare = pd.read_pickle('/content/drive/My Drive/project/seoul_contest/복지시설(대중교통,도서관,박물관,미술관,공원)행정동별매칭.pkl')
med = pd.read_pickle('/content/drive/My Drive/project/seoul_contest/건축대장_의료시설.pickle')


### 주차장 + 의료시설 + 범죄율 + 초등인구 합치기 !

In [17]:
med.drop(['시군구명','행정동명'],axis=1,inplace=True)

In [32]:
med

15     1111051500
0      1111053000
1      1111054000
2      1111055000
3      1111056000
          ...    
417    1174065000
418    1174066000
423    1174068500
419    1174069000
420    1174070000
Name: 행정동코드, Length: 425, dtype: int64

In [27]:
crime = crime['행정동코드'].sort_values()

In [34]:
# med = med[med['행정동코드'].sort_values()]
med.sort_values('행정동코드',inplace=True)
crime.sort_values('행정동코드',inplace=True)



In [58]:
crime[crime['행정동']=='신사동']

행정동코드     구+동 시군구명 행정동(제)  ...       동(발생)       동(검거)    동(검거율)   인구대비범죄율
333  1162068500  관악구신사동  관악구    신사동  ...  268.052107  191.681406  0.715090  0.010319
359  1168051000  강남구신사동  강남구    신사동  ...  230.958953  160.286272  0.694003  0.013373

[2 rows x 17 columns]

In [55]:
age = pd.read_csv('/content/drive/My Drive/project/seoul_contest/인구 (1).csv',encoding='cp949')

In [64]:
med['구+동'] = med['시군구명']+med['행정동명']
crime['구+동2'] = crime['시군구명']+crime['행정동(제)']

In [65]:
med

행정동코드 시군구명   행정동명  의료시설_개수  의료시설_건축면적(㎡)  의료시설_건폐율(%)       구+동
15   1111051500  종로구  청운효자동      1.0      0.000000     0.000000  종로구청운효자동
0    1111053000  종로구    사직동      2.0    104.130000    39.368600    종로구사직동
1    1111054000  종로구    삼청동      0.0      0.000000     0.000000    종로구삼청동
2    1111055000  종로구    부암동      0.0      0.000000     0.000000    종로구부암동
3    1111056000  종로구    평창동      1.0   2012.800000    44.220000    종로구평창동
..          ...  ...    ...      ...           ...          ...       ...
417  1174065000  강동구  성내제2동      7.0    523.218571    46.427143  강동구성내제2동
418  1174066000  강동구  성내제3동      7.0    523.218571    46.427143  강동구성내제3동
423  1174068500  강동구     길동      9.0    734.660000    51.418478     강동구길동
419  1174069000  강동구  둔촌제1동      8.0   3078.785000     0.000000  강동구둔촌제1동
420  1174070000  강동구  둔촌제2동      8.0   3078.785000     0.000000  강동구둔촌제2동

[425 rows x 7 columns]

In [59]:
age['구+동'] = age['자치구']+age['동']
age

자치구     동      합      구+동
0    종로구   사직동    715   종로구사직동
1    종로구   삼청동    204   종로구삼청동
2    종로구   부암동    805   종로구부암동
3    종로구   평창동  1,425   종로구평창동
4    종로구   무악동    920   종로구무악동
..   ...   ...    ...      ...
420  강동구  둔촌1동     11  강동구둔촌1동
421  강동구  둔촌2동  2,060  강동구둔촌2동
422  강동구  암사1동  2,468  강동구암사1동
423  강동구  천호2동  1,755  강동구천호2동
424  강동구    길동  2,798    강동구길동

[425 rows x 4 columns]

In [66]:
df = pd.merge(med.set_index('구+동'),crime.set_index('구+동2'),left_index=True,right_index=True)
# all_m = pd.merge(w.set_index('행정동코드'),whole.set_index('행정동코드'),how='outer',left_index=True,right_index=True)
df

행정동코드_x 시군구명_x   행정동명  ...       동(검거)    동(검거율)   인구대비범죄율
종로구청운효자동  1111051500    종로구  청운효자동  ...  329.703682  1.070463  0.023960
종로구사직동    1111053000    종로구    사직동  ...  251.015942  1.070463  0.023960
종로구삼청동    1111054000    종로구    삼청동  ...   76.251190  1.070463  0.023960
종로구부암동    1111055000    종로구    부암동  ...  267.276707  1.070463  0.023960
종로구평창동    1111056000    종로구    평창동  ...  479.513033  1.070463  0.023960
...              ...    ...    ...  ...         ...       ...       ...
강동구성내제2동  1174065000    강동구  성내제2동  ...  160.674168  0.736173  0.008697
강동구성내제3동  1174066000    강동구  성내제3동  ...  152.530338  0.736173  0.008697
강동구길동     1174068500    강동구     길동  ...  300.591814  0.736173  0.008697
강동구둔촌제1동  1174069000    강동구  둔촌제1동  ...    1.754253  0.736173  0.008697
강동구둔촌제2동  1174070000    강동구  둔촌제2동  ...  174.733797  0.736173  0.008697

[425 rows x 23 columns]

In [67]:
df2 = pd.merge(df.set_index('구+동'),age.set_index('구+동'),left_index=True,right_index=True)
df2

행정동코드_x 시군구명_x   행정동명  의료시설_개수  ...   인구대비범죄율  자치구      동      합
구+동                                          ...                             
종로구청운효자동  1111051500    종로구  청운효자동      1.0  ...  0.023960  종로구  청운효자동  1,115
종로구사직동    1111053000    종로구    사직동      2.0  ...  0.023960  종로구    사직동    715
종로구삼청동    1111054000    종로구    삼청동      0.0  ...  0.023960  종로구    삼청동    204
종로구부암동    1111055000    종로구    부암동      0.0  ...  0.023960  종로구    부암동    805
종로구평창동    1111056000    종로구    평창동      1.0  ...  0.023960  종로구    평창동  1,425
...              ...    ...    ...      ...  ...       ...  ...    ...    ...
강동구성내2동   1174065000    강동구  성내제2동      7.0  ...  0.008697  강동구   성내2동  1,142
강동구성내3동   1174066000    강동구  성내제3동      7.0  ...  0.008697  강동구   성내3동  1,440
강동구길동     1174068500    강동구     길동      9.0  ...  0.008697  강동구     길동  2,798
강동구둔촌1동   1174069000    강동구  둔촌제1동      8.0  ...  0.008697  강동구   둔촌1동     11
강동구둔촌2동   1174070000    강동구  둔촌제2동      8.0  ...  0.008697  강동구   둔촌2동  2,060

[425 rows x 25 columns]

In [70]:
df2.reset_index(inplace=True)
df2.columns

Index(['level_0', 'index', '구+동', '행정동코드_x', '시군구명_x', '행정동명', '의료시설_개수',
       '의료시설_건축면적(㎡)', '의료시설_건폐율(%)', '행정동코드_y', '시군구명_y', '행정동(제)', '행정동',
       '인구(동)', '인구(구)', '합계(개소)', '합계(면수)', '동비율', '합계(발생)', '합계(검거)', '검거율',
       '동(발생)', '동(검거)', '동(검거율)', '인구대비범죄율', '자치구', '동', '합'],
      dtype='object')

In [72]:
df2.drop(['level_0', 'index', '시군구명_x', '행정동명', '행정동코드_y'],axis='columns', inplace=True)

In [75]:
df2.drop(['구+동','시군구명_y','행정동(제)','동비율'],axis='columns', inplace=True)

In [87]:
col = [ '의료시설_개수', '의료시설_건축면적(㎡)', '의료시설_건폐율(%)',  '합계(개소)', '합계(면수)', '합계(발생)', '합계(검거)', '검거율', '동(발생)','동(검거)', '동(검거율)', '인구대비범죄율','인구(동)','인구(구)','행정동코드_x' ,'자치구','행정동', '동', '합']

In [88]:
df2 = df2[col]

In [89]:
df2

의료시설_개수  의료시설_건축면적(㎡)  의료시설_건폐율(%) 합계(개소)  ...  자치구    행정동      동      합
0        1.0      0.000000     0.000000    668  ...  종로구  청운효자동  청운효자동  1,115
1        2.0    104.130000    39.368600    360  ...  종로구    사직동    사직동    715
2        0.0      0.000000     0.000000    169  ...  종로구    삼청동    삼청동    204
3        0.0      0.000000     0.000000    596  ...  종로구    부암동    부암동    805
4        1.0   2012.800000    44.220000  1,600  ...  종로구    평창동    평창동  1,425
..       ...           ...          ...    ...  ...  ...    ...    ...    ...
420      7.0    523.218571    46.427143  1,935  ...  강동구   성내2동   성내2동  1,142
421      7.0    523.218571    46.427143  1,431  ...  강동구   성내3동   성내3동  1,440
422      9.0    734.660000    51.418478  2,032  ...  강동구     길동     길동  2,798
423      8.0   3078.785000     0.000000     21  ...  강동구   둔촌1동   둔촌1동     11
424      8.0   3078.785000     0.000000  1,116  ...  강동구   둔촌2동   둔촌2동  2,060

[425 rows x 19 columns]

In [102]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 425 entries, 0 to 424
Data columns (total 19 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   의료시설_개수       425 non-null    float64
 1   의료시설_건축면적(㎡)  425 non-null    float64
 2   의료시설_건폐율(%)   425 non-null    float64
 3   합계(개소)        423 non-null    object 
 4   합계(면수)        423 non-null    object 
 5   합계(발생)        425 non-null    float64
 6   합계(검거)        425 non-null    float64
 7   검거율           425 non-null    float64
 8   동(발생)         425 non-null    float64
 9   동(검거)         425 non-null    float64
 10  동(검거율)        425 non-null    float64
 11  인구대비범죄율       425 non-null    float64
 12  인구(동)         425 non-null    object 
 13  인구(구)         425 non-null    object 
 14  행정동코드_x       425 non-null    int64  
 15  자치구           425 non-null    object 
 16  행정동           425 non-null    object 
 17  동             425 non-null    object 
 18  합             425 non-null    

In [103]:
# df2.iloc[:,-1].str.replace(',', '').astype('int64')
df2['합계(개소)']= df2['합계(개소)'].str.replace(',', '').astype('float64')
df2['합계(면수)']= df2['합계(면수)'].str.replace(',', '').astype('float64')
df2['합']= df2['합'].str.replace(',', '').astype('float64')
df2['인구(동)']= df2['인구(동)'].str.replace(',', '').astype('float64')


- 여의도동 / 여의동 표기 차이로 오류난 data 있음
- 주차장 데이터는 '항동' NaN 처리
- RandomForest 하기 위해서 null 값 떨굼

In [112]:
df2.dropna(inplace=True)

In [130]:
df2.to_pickle('/content/drive/My Drive/project/seoul_contest/의료주차범죄합친것.pkl')

## 예측 모델링

In [114]:
X,y_y,y_all = df2.iloc[:,:12], df2.iloc[:,-1],df2.iloc[:,12]

In [115]:
scaler = MinMaxScaler()
X_scaled = scaler.fit_transform(X)

### target 동 전체 인구

In [116]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y_y, random_state=42, test_size=0.2)

In [117]:
forest_reg = RandomForestRegressor(1000, max_depth=2, random_state=10, n_jobs=-1)

train score ( 전체 인구)

In [120]:
forest_reg.fit(X_train, y_train)
forest_reg.score(X_train, y_train)

0.49817688693061457

test score ( 전체 인구)

In [121]:
forest_reg.score(X_test, y_test)
forest_reg.score(X_test, y_test)


0.45049000285433843

In [131]:
from sklearn.metrics import mean_squared_error
np.sqrt(mean_squared_error(forest_reg.predict(X_train), y_train))

22672.93307552181

#### target : 초딩

In [123]:
X_train_, X_test_, y_train_, y_test_ = train_test_split(X_scaled, y_all, random_state=42, test_size=0.2)

train score  (y:초등)

In [124]:
forest_reg.fit(X_train_, y_train_)
forest_reg.score(X_train_, y_train_)

0.6887945599773589

test score  (y:초등)

In [125]:
forest_reg.score(X_test_, y_test_)
forest_reg.score(X_test_, y_test_)


0.6917544959291535

### X : 범죄율/ 주차장 제외한 모두 

In [132]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [133]:
total=pd.read_csv('/content/drive/My Drive/project/seoul_contest/total_합치기.csv')
total

Unnamed: 0       행정동코드 시군구명   행정동명  ...  5~9세  10~14세  5~14세     아동 비율
0             0  1111051500  종로구  청운효자동  ...  1679    1178   2857  0.101799
1             1  1111053000  종로구    사직동  ...  1581    1750   3331  0.099424
2             2  1111054000  종로구    삼청동  ...   827     809   1636  0.059029
3             3  1111055000  종로구    부암동  ...   261     193    454  0.030142
4             4  1111056000  종로구    평창동  ...   398     336    734  0.029561
..          ...         ...  ...    ...  ...   ...     ...    ...       ...
420         420  1174065000  강동구  성내제2동  ...   386     350    736  0.070858
421         421  1174066000  강동구  성내제3동  ...   584     590   1174  0.063964
422         422  1174068500  강동구     길동  ...   365     310    675  0.043771
423         423  1174069000  강동구  둔촌제1동  ...   765     955   1720  0.068439
424         424  1174070000  강동구  둔촌제2동  ...  1314    1014   2328  0.066529

[425 rows x 30 columns]

In [145]:
all = pd.merge(total.set_index('행정동코드'),df2.set_index('행정동코드_x'),left_index=True,right_index=True)
col = ['Unnamed: 0', '시군구명', '행정동명', '공동주택_개수', '공동주택_건축면적(㎡)', '공동주택_건폐율(%)',
       '공동주택_연면적(㎡)', '공동주택_용적률(%)', '공동주택_세대수', '공동주택_가구수', '단독주택_개수',
       '단독주택_건축면적(㎡)', '단독주택_건폐율(%)', '단독주택_가구수', '공장_개수', '공장_건축면적(㎡)',
       '동별총학원', '동별입시학원수', '버스정류장', '도서관', '박물관/미술관', '공원', '지하철역', '병원수',
       '유통개수', '의료시설_개수', '의료시설_건축면적(㎡)','의료시설_건폐율(%)', '합계(개소)', '합계(면수)', '합계(발생)', '합계(검거)', '검거율', '동(발생)',
       '동(검거)', '동(검거율)', '인구대비범죄율',  '5~9세', '10~14세', '5~14세', '아동 비율','인구(동)', '인구(구)', '자치구', '행정동', '동', '합']
all = all[col]

In [146]:
all

Unnamed: 0 시군구명   행정동명  공동주택_개수  ...  자치구    행정동      동       합
1111051500           0  종로구  청운효자동      379  ...  종로구  청운효자동  청운효자동  1115.0
1111053000           1  종로구    사직동       98  ...  종로구    사직동    사직동   715.0
1111054000           2  종로구    삼청동       14  ...  종로구    삼청동    삼청동   204.0
1111055000           3  종로구    부암동      246  ...  종로구    부암동    부암동   805.0
1111056000           4  종로구    평창동      454  ...  종로구    평창동    평창동  1425.0
...                ...  ...    ...      ...  ...  ...    ...    ...     ...
1174065000         420  강동구  성내제2동     1247  ...  강동구   성내2동   성내2동  1142.0
1174066000         421  강동구  성내제3동     1247  ...  강동구   성내3동   성내3동  1440.0
1174068500         422  강동구     길동      683  ...  강동구     길동     길동  2798.0
1174069000         423  강동구  둔촌제1동      359  ...  강동구   둔촌1동   둔촌1동    11.0
1174070000         424  강동구  둔촌제2동      359  ...  강동구   둔촌2동   둔촌2동  2060.0

[423 rows x 47 columns]

In [149]:
all.iloc[:,3:-10]

공동주택_개수  공동주택_건축면적(㎡)  공동주택_건폐율(%)  ...       동(검거)    동(검거율)   인구대비범죄율
1111051500      379    125.269681    52.555743  ...  329.703682  1.070463  0.023960
1111053000       98    348.211020    44.991429  ...  251.015942  1.070463  0.023960
1111054000       14    203.559286    34.089286  ...   76.251190  1.070463  0.023960
1111055000      246    124.504472    24.580998  ...  267.276707  1.070463  0.023960
1111056000      454    389.523883    20.485875  ...  479.513033  1.070463  0.023960
...             ...           ...          ...  ...         ...       ...       ...
1174065000     1247    193.674798    46.141739  ...  160.674168  0.736173  0.008697
1174066000     1247    193.674798    46.141739  ...  152.530338  0.736173  0.008697
1174068500      683    232.939603    41.283617  ...  300.591814  0.736173  0.008697
1174069000      359    231.635373    36.682957  ...    1.754253  0.736173  0.008697
1174070000      359    231.635373    36.682957  ...  174.733797  0.736173  0.008697

[423 rows x 34 columns]

In [150]:
X_all,Y = all.iloc[:,3:-10], all.iloc[:,-8]
scaler = MinMaxScaler()
X_scaled_all = scaler.fit_transform(X_all)

In [151]:
x_train_, x_test_, Y_train_, Y_test_ = train_test_split(X_scaled_all, Y, random_state=42, test_size=0.2)

train score (y : 5-14)

In [152]:
forest_reg.fit(x_train_, Y_train_)
forest_reg.score(x_train_, Y_train_)

0.1771103068796127

test score (y : 5-14)

In [153]:
forest_reg.fit(x_test_, Y_test_)
forest_reg.score(x_test_, Y_test_)

0.4632453018108804

주차장,의료시설,범죄율 은 0.69 까지 올라오는데...

전체 데이터는 왜 0.46 밖에 안되는지 모르겠네 ? ???
